File: quicksort.py
import random

def quick_sort(array, start, end):
    if start >= end: # 원소가 1개인 경우 종료
        return
    pivot = start # 피벗은 첫 번째 원소
    left = start + 1
    right = end
    while(left <= right):
        # 피벗보다 큰 데이터를 찾을 때까지 반복 
        while(left <= end and array[left] <= array[pivot]):
            left += 1
        # 피벗보다 작은 데이터를 찾을 때까지 반복
        while(right > start and array[right] >= array[pivot]):
            right -= 1
        if(left > right): # 엇갈렸다면 작은 데이터와 피벗을 교체
            array[right], array[pivot] = array[pivot], array[right]
        else: # 엇갈리지 않았다면 작은 데이터와 큰 데이터를 교체
            array[left], array[right] = array[right], array[left]
    # 분할 이후 왼쪽 부분과 오른쪽 부분에서 각각 정렬 수행
    quick_sort(array, start, right - 1)
    quick_sort(array, right + 1, end)


def checksort(a):
    Sorted = True
    n = len(a)
    for i in range(n-1):
        if a[i] > a[i+1]:
            Sorted = False
        if not Sorted:
            break
    if Sorted: print('정렬 완료.')
    else: print('정렬 오류 발생.')
        

num_list = []

for i in range(20):
  num_list.append(random.randint(0, 100)) 

print("before sort = ", num_list)
quick_sort(num_list, 0, len(num_list) - 1)

print("after sort = ",num_list)
checksort(num_list)

File: src/algorithms/sorting/js/quick-sort.js
function swap(arr, a, b) {
  const temp = arr[a];
  arr[a] = arr[b];
  arr[b] = temp;
}

function quickSort(arr, start, end) {
  if (end <= start) return;
  const pivot = start;
  let left = start + 1;
  let right = end;

  // 모든 요소를 비교하교 right, left가 어긋날 때 까지
  while (right >= left) {

    // left가 end보다 작거나 같고, left 위치의 요소가 피벗보다 작거나 같다면 left + 1
    while (left <= end && arr[left] <= arr[pivot]) {
      left++;
    }
    // right가 start보다 크고, right 위치의 요소가 피벗보다 크다면 right - 1
    while (right > start && arr[right] > arr[pivot]) {
      right--;
    }
    // 모든 요소를 비교했다면 right과 피벗 교환
    if (left > right) {
      swap(arr, pivot, right);
    // 아니면 left, right 요소 교환
    } else {
      swap(arr, left, right);
    }
    console.log(arr, left, right)
  }
  // 재귀
  quickSort(arr, start, right - 1);
  quickSort(arr, left, end);
}

const arr = [
  2,2,3,2,4,5,6,7,8,2
]
// for (let i = 0; i < 10; i++) {
//   arr.push(Math.floor(Math.random() * 100));
// }


quickSort(arr, 0, arr.length - 1);

